### PR TITLE
Support sliding window and per-head sinks in Triton varlen prefill attention

### PR DESCRIPTION
**Pull request by Liu-congo**
## Motivation
#37983 

## Modifications

Kernel (context_attention_fwd / _fwd_kernel):
- Add window_size=(left, right) with flash-attn convention (-1 =
  unbounded): distance mask composed with the sequence-boundary/causal
  masks, plus KV-block loop tightening on both ends so small windows skip
  out-of-range blocks.
- Add optional per-q-head sinks: a value-less virtual logit appended to
  the softmax denominator. Implemented by seeding the online-softmax
  state with (m = s, l = 1), i.e. "process the sink first"; every later
  rescale carries exp(s - m) in the denominator while the loop body stays
  untouched. Semantics match the FA3 vision path (raw additive logit, per
  q-head, not scaled by softmax_scale).
- Guard the windowed path against fully-masked rows (a row's window can
  miss a whole KV block): clamp all -inf m_ij to a finite value and hold
  the p/acc scales at 0 while l == 0, otherwise exp(-inf - (-inf)) = NaN
  poisons every later block.
- WINDOW_LEFT / WINDOW_RIGHT / HAS_SINKS are constexpr, so non-windowed
  sink-free callers compile to the exact previous code: zero regression
  for existing users. New params are keyword-with-defaults; positional
  callers (extend path) are unaffected.
- Validate sinks (one logit per q head, contiguous, same device): the
  kernel reads it as a raw pointer, so a bad tensor would otherwise fail
  silently or crash.
Backend:
- VisionTritonAttention.forward now forwards window_size / s_aux to the
  kernel.
Tests (test/registered/attention/test_triton_attention_window_sink.py):
model-independent oracle built from a PyTorch reference, anchored by a
hand-computed example; covers window/sinks/GQA/varlen, the multi-KV-block
NaN trap with small and one-sided windows, causal+window(+sinks), bf16
sinks (the model-side parameter dtype), the issue reproducer at the
backend layer, and an optional FA3 cross-check on Hopper.

## Accuracy Tests

All numbers from the new test suite (13 passed + 9 subtests, RTX 5090 /
SM120, torch 2.13+cu130, bf16 inputs with fp32 accumulation reference):

- **Kernel vs mathematical reference**: max |diff| ≈ 0.002–0.004 across
  window/sink/GQA/varlen/causal combinations (tolerance atol=rtol=2e-2).
- **Issue reproducer (backend layer)**: before the fix, output was
  bit-identical regardless of `window_size`/`s_aux` (kwargs dropped) and
  matched full-attention-without-sink semantics with ~75% relative error;
  after the fix it matches the reference to ~0.2% relative error.
- **FA3 parity**: sink semantics (raw additive per-q-head logit, not
  scaled by softmax_scale) verified against the FA3 cute kernel source;
  the numeric cross-check (`test_fa3_cross_check_reference`) runs on
  Hopper CI and skips elsewhere.
- **Edge cases**: extreme sink logits (±88, exp args are ≤ 0 by
  construction so no overflow), fully-masked KV blocks with small windows
  (no NaN), padding rows, empty-sequence batches.
- **Regressions**: `test_vision_max_seqlen.py` 7 passed;
  `test_triton_attention_kernels.py` unchanged except pre-existing
  `OutOfResources` failures on this 99 KB-smem consumer GPU (identical
  with the pre-change kernel baseline).

## Speed Tests and Profiling

Microbenchmark, original kernel loaded from the pre-change source vs new
kernel (RTX 5090, bf16, Hq=8, D=64, CUDA-event timed, 50 iters):

| workload | case | time | vs baseline |
|---|---|---|---|
| 4 × 1369 tokens (high-res vision) | full attn, **original kernel** | 173.0 µs | 1.00× |
| | full attn, new (no window/sink) | 173.0 µs | **1.00× (zero regression)** |
| | full attn, new + sinks | 173.1 µs | 1.00× (sink ≈ free) |
| | window(64,64) | 46.3 µs | **3.73×** |
| | window(64,64) + sinks | 46.2 µs | 3.74× |
| mixed varlen, 2693 tokens | full attn, original vs new | 104.5 / 104.3 µs | 1.00× |
| | window(64,64) + sinks | 35.4 µs | 2.95× |

Non-windowed callers hit the identical compiled code (constexpr
specialization); sinks add one scalar load per program (outside the KV
loop); windowed calls skip out-of-range KV blocks and are strictly faster
than the full attention they previously (incorrectly) fell back to.

## Checklist

- [] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34503082795](https://github.com/sgl-project/sglang/actions/runs/34503082795)
Latest PR Test (Extra): :x: [Run #34503082573](https://github.com/sgl-project/sglang/actions/runs/34503082573)
Latest PR Test (AMD ROCm 10): :x: [Run #34503082718](https://github.com/sgl-project/sglang/actions/runs/34503082718)

### CODE REVIEW

**Comment by Liu-congo** [comment]
@Fridge003 hello，could you help with a code review?

**Comment by Liu-congo** [comment]
@yhyang201 @DarkSharpness hello, can you help review this pr

**Comment by yeungtuzi** [comment]
Independent verification on a different architecture than the author's (A100, **SM80**, torch 2.13.0+cu130 / triton 3.7.1, bf16 inputs against an fp32 PyTorch reference): the patched kernel matches the reference to **0.26-0.27% relative error** across full attention, `window(64,64)`, `window+sink`, `causal+window+sink`, GQA, varlen, and one-sided-window cases. The pre-fix kernel is **65-95% off** on every window/sink case, i.e. it silently computes full attention. Full-attention output is **bit-identical** between the old and new kernels (max abs diff 0.0), which independently confirms the "zero regression for existing callers" claim beyond the RTX 5090 / SM120 numbers already posted.

One CI blocker remains: the `Lint` job fails on a single taxonomy rule - registered kernel tests must live under `test/registered/kernels/{ops,benchmark}/<group>/` (see `scripts/lint/check_registered_tests.py`). Moving `test/registered/attention/test_triton_attention_window_sink.py` to `test/registered/kernels/ops/attention/test_triton_attention_window_sink.py` resolves the only failing hook; all other pre-commit hooks pass. Note the PR-test workflows are currently skipped at `pr-gate`, so the new suite has not actually run in CI yet.

Ping @BBuf (Kernel merge oncall) and @Fridge003 (attention) for review. This fixes silent wrongness for MiMo-style ViTs whose ViT blocks use local windows plus learned sinks (sgl-project/sglang#37983), not a new feature.

**Comment by ratulsarna** [comment]
For MiMo-V2.6 vision, the checkpoint uses a different sink definition from the value-less virtual logit in this PR.

The [pinned checkpoint reference](https://huggingface.co/XiaomiMiMo/MiMo-V2.6-Flash-RL/blob/3b38d063180c3e4aed9691fdc735f3d10b266ee4/modeling_mimo_v2.py#L712-L723) adds `sinks[h]` to the score for the first **real** key, after QK scaling. The local-window mask still applies to that key, and its actual `V[0]` contributes when it is in the window:

```text
softmax(QK^T * scale + window_mask + sinks[h] * (key_index == 0)) @ V
```

A value-less extra logit changes both the normalization and the contribution from `V[0]`. Agreement with FA3's virtual-sink definition does not establish agreement with this checkpoint. [vLLM's merged MiMo fix](https://github.com/vllm-project/vllm/pull/49815) also applies the bias to the first real key.

We tested a MiMo-only implementation on two GB10/SM121 GPUs with BF16 Q/K/V, 16 query heads, 4 KV heads and head dimension 64. All 12 cases on each GPU matched an FP32 SDPA reference within max absolute error 0.01414 and RMSE 0.00109. The cases cover variable lengths, strided tensors, inclusive windows, masked intermediate tiles, and bitwise preservation of the original full/causal paths. Deliberately substituting a value-less sink differs from the checkpoint reference by more than 1 in representative cases.

With speculation disabled, exact synthetic FHD and 4K cards that previously produced incorrect IDs/colors returned the expected answers after the attention correction. The reproducer uses one image per request to isolate attention from multi-image scheduling.

The implementation, GPU receipts and synthetic-image reproducer are in [MiaAI recipe PR #2](https://github.com/MiaAI-Lab/MiMo-V2.6-Flash-2x-DGX-Sparks/pull/2). That PR also includes a separate DFlash image-forwarding correction.

Could this support explicit real-key-zero semantics for MiMo while retaining the virtual-sink behavior for callers that require it?

**Comment by yeungtuzi** [comment]
Thanks for the pointer to the checkpoint reference — that is a useful distinction, and I want to be precise about what my verification did and did not cover.

My comment addresses whether the kernel computes the semantics it declares, on an architecture other than the author's. The oracle I used was my own PyTorch reference implementing the virtual-sink formulation, so the ~0.26% agreement, the bitwise identity of the full-attention path, and the pre-fix 65–95% divergence are statements about numerical correctness, cross-architecture reproducibility, and the no-regression claim. They are not a claim that virtual-sink semantics match MiMo. Where the oracle defines the semantics, it cannot also validate the choice of semantics — so I agree the two questions are separate, and I have no evidence to offer against the checkpoint behavior you describe.

What would make the MiMo-consistency claim actionable here, rather than a property of an out-of-tree implementation:

- A minimal registered test in this PR whose oracle is derived from the pinned checkpoint reference (key-0 bias, with `V[0]` contributing when in-window), rather than from a value-less logit. That is the piece that would turn the current measurements into in-repo evidence, in the same way your reproducer anchors the other side of the question.
- An explicit decision from the author on whether to support both modes, given the vLLM precedent (#49815 adds `sinks_bias_key0` alongside the existing null-logit path). If both are supported, it would help to state which one the MiMo path selects and why.

One note on scope: the reported GB10 figures are against your own implementation and reference, so I read them as evidence that the two formulations differ, which is consistent with the vLLM measurements — not yet as independent confirmation of the model-side semantics. Packaging the MiMo case as a test against the checkpoint reference would settle that directly, and it would also be cheap insurance against the window and sink paths drifting apart later.

Separately, the CI blocker from my earlier comment is still open and is orthogonal to this discussion: the `Lint` job fails because registered kernel tests must live under `test/registered/kernels/{ops,benchmark}/<group>/` (`scripts/lint/check_registered_tests.py`). Moving `test/registered/attention/test_triton_attention_window_sink.py` to `test/registered/kernels/ops/attention/` clears the only failing hook.
